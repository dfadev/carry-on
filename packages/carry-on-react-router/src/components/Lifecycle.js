import React from "react";

class Lifecycle extends React.Component {
  componentDidMount() {
    const { onMount } = this.props;
    if (onMount) onMount.call(this, this);
  }

  componentDidUpdate(prevProps) {
    const { onUpdate } = this.props;
    if (onUpdate) onUpdate.call(this, this, prevProps);
  }

  componentWillUnmount() {
    const { onUnmount } = this.props;
    if (onUnmount) onUnmount.call(this, this);
  }

  render() {
    return null;
  }
}

export default Lifecycle;
