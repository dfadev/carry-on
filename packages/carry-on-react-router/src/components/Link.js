import React from "react";
import { carryOn } from "carry-on-react";
import { getIn } from "carry-on-utils";

const Link = carryOn(
  { id: "Link" },
  (
    {
      history = "app.history",
      innerRef,
      replace,
      to,
      onClick,
      target,
      force,
      ...rest
    },
    state
  ) => {
    const { handleClick, getHref } = getIn(state, history);

    return (
      <a
        {...rest}
        onClick={handleClick({ replace, to, onClick, target, force })}
        href={getHref(to)}
        ref={innerRef}
      />
    );
  }
);

export default Link;
