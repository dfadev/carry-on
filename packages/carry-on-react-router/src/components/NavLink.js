import React from "react";
import { carryOn } from "carry-on-react";
import { getIn } from "carry-on-utils";

function joinClassnames(...classnames) {
  return classnames.filter(i => i).join(" ");
}

const escapedRE = /([.+*?=^!:${}()[\]|/\\])/g;

const NavLink = carryOn(
  { id: "NavLink" },
  (
    {
      history = "app.history",
      "aria-current": ariaCurrent = "page",
      activeClassName = "active",
      activeStyle,
      className: classNameProp,
      exact,
      isActive: isActiveProp,
      location: loc,
      style: styleProp,
      strict,
      sensitive,
      to,
      replace,
      onClick,
      target,
      innerRef,
      ...rest
    },
    state
  ) => {
    const hist = getIn(state, history);
    const location = loc || hist.location;
    const path = typeof to === "object" ? to.pathname : to;
    const escapedPath = path && path.replace(escapedRE, "\\$1");
    const match = hist.matchPath({
      path: escapedPath,
      exact,
      strict,
      sensitive
    });
    const isActive = !!(isActiveProp ? isActiveProp(match, location) : match);
    const className = isActive
      ? joinClassnames(classNameProp, activeClassName)
      : classNameProp;
    const style = isActive ? { ...styleProp, ...activeStyle } : styleProp;
    const { handleClick, getHref } = hist;

    /* eslint-disable jsx-a11y/anchor-has-content */
    return (
      <a
        {...rest}
        onClick={handleClick({ replace, to, onClick, target })}
        href={getHref(to)}
        ref={innerRef}
        style={style}
        className={className}
        aria-current={(isActive && ariaCurrent) || null}
      />
    );
  }
);

export default NavLink;
