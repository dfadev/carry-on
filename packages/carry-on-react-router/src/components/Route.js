import React from "react";
import { carryOn } from "carry-on-react";
import { getIn } from "carry-on-utils";

function isEmptyChildren(children) {
  return React.Children.count(children) === 0;
}

const Route = carryOn((props, state) => {
  const {
    history = "app.history",
    location: loc,
    computedMatch,
    path,
    children: childs,
    component,
    render
  } = props;

  const hist = getIn(state, history);
  const location = loc || hist.location;

  let match;
  if (computedMatch) match = computedMatch;
  else if (path) match = hist.matchPath(props);
  //else match = hist.match; // ???

  const renderProps = { location, match };

  let children = childs;
  if (Array.isArray(children) && children.length === 0)
    children = null;

  if (typeof children === "function") {
    children = children(renderProps);
    if (children === undefined) children = null;
  }

  if (children && !isEmptyChildren(children)) return children;
  if (match)
    if (component) return React.createElement(component, renderProps);
    else if (render) return render(renderProps);

  return null;
});

Route.displayName = "Route";

export default Route;
