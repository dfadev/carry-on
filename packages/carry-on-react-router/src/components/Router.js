/* eslint-disable max-classes-per-file */
import React, { useMemo } from "react";
import {
  withStore,
  State,
  Register,
  Middleware,
  OnUnmount,
  Render
} from "carry-on-react";
import {
  createBrowserHistory,
  createHashHistory,
  createMemoryHistory
} from "history";
import createStaticHistory from "../createStaticHistory";
import router from "../router";

function GenericRouter(createHistory = createBrowserHistory) {
  return withStore(({ children = null, path = "app.history", ...props }) => {
    const rtr = useMemo(() => router(createHistory(props), path), [path]);

    return (
      <State {...props}>
        <Register>{rtr.state}</Register>
        <Middleware>{rtr.middleware}</Middleware>
        <OnUnmount>
          {state =>
            state &&
            state.app &&
            state.app.history &&
            state.app.history.unlisten &&
            state.app.history.unlisten()
          }
        </OnUnmount>
        <Render>{() => children || null}</Render>
      </State>
    );
  });
}

export const Router = GenericRouter(props => props.history);
export const MemoryRouter = GenericRouter(createMemoryHistory);
export const BrowserRouter = GenericRouter(createBrowserHistory);
export const HashRouter = GenericRouter(createHashHistory);
export const StaticRouter = GenericRouter(createStaticHistory);
