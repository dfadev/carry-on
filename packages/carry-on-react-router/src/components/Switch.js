import React from "react";
import { carryOn } from "carry-on-react";
import { getIn } from "carry-on-utils";
import matchPath from "../matchPath";

const Switch = carryOn(
  { id: "Switch" },
  ({ location: loc, children, history = "app.history" }, state) => {
    const hist = getIn(state, history);
    const location = loc || hist.location;

    let element;
    let match;

    React.Children.forEach(children, child => {
      if (match == null && React.isValidElement(child)) {
        element = child;
        match = child.props.path
          ? matchPath(location.pathname, { ...child.props })
          : hist.match;
      }
    });

    return match
      ? React.cloneElement(element, { location, computedMatch: match })
      : null;
  }
);

export default Switch;
