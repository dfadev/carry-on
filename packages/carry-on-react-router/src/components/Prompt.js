import React from "react";
import { carryOn } from "carry-on-react";
import { getIn } from "carry-on-utils";
import Lifecycle from "./Lifecycle";

/**
 * The public API for prompting the user before navigating away from a screen.
 */
const Prompt = carryOn(
  { id: "Prompt" },
  ({ history = "app.history", message, when = true }, state) => {
    const hist = getIn(state, history);

    if (!when || hist.staticContext) return null;

    const method = hist.block;

    return (
      <Lifecycle
        onMount={self => {
          self.release = method(message);
        }}
        onUpdate={(self, prevProps) => {
          if (prevProps.message !== message) {
            self.release();
            self.release = method(message);
          }
        }}
        onUnmount={self => {
          self.release();
        }}
        message={message}
      />
    );
  }
);

export default Prompt;
