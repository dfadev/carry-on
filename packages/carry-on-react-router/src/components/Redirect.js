import React from "react";
import { createLocation, locationsAreEqual } from "history";
import { carryOn } from "carry-on-react";
import { getIn } from "carry-on-utils";

import Lifecycle from "./Lifecycle";
import generatePath from "../generatePath";

const Redirect = carryOn(
  { id: "Redirect" },
  ({ history = "app.history", computedMatch, to, push = false }, state) => {
    const hist = getIn(state, history);
    const method = push ? hist.push : hist.replace;

    /* eslint-disable no-nested-ternary */
    const location = createLocation(
      computedMatch
        ? typeof to === "string"
          ? generatePath(to, computedMatch.params)
          : {
              ...to,
              pathname: generatePath(to.pathname, computedMatch.params)
            }
        : to
    );

    // When rendering in a static context, set the new location immediately.
    if (hist.staticContext) {
      method(location);
      return null;
    }

    return (
      <Lifecycle
        onMount={() => {
          method(location);
        }}
        onUpdate={(_, prevProps) => {
          if (!locationsAreEqual(prevProps.to, location)) {
            method(location);
          }
        }}
        to={to}
      />
    );
  }
);

export default Redirect;
