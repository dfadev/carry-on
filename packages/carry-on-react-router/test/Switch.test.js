/**
 * @jest-environment jsdom
 */
import React from "react";
import { render } from "@testing-library/react";
import { initStores } from "carry-on-store";
import { MemoryRouter } from "../src/components/Router";
import Route from "../src/components/Route";
import Redirect from "../src/components/Redirect";
import Switch from "../src/components/Switch";

describe("A <Switch>", () => {
  afterEach(() => {
    initStores();
  });

  describe("without a <Router>", () => {
    it("throws an error", () => {
      jest.spyOn(console, "error").mockImplementation(() => {});

      expect(() => {
        render(<Switch />);
      }).toThrow();
    });
  });

  it("renders the first <Route> that matches the URL", () => {
    expect(
      render(
        <MemoryRouter initialEntries={["/one"]}>
          <Switch>
            <Route path="/one" render={() => <h1>one</h1>} />
            <Route path="/two" render={() => <h1>two</h1>} />
          </Switch>
        </MemoryRouter>
      ).asFragment()
    ).toMatchSnapshot();
  });

  it("does not render a second <Route> that also matches the URL", () => {
    expect(
      render(
        <MemoryRouter initialEntries={["/one"]}>
          <Switch>
            <Route path="/one" render={() => <h1>one</h1>} />
            <Route path="/one" render={() => <h1>two</h1>} />
          </Switch>
        </MemoryRouter>
      ).asFragment()
    ).toMatchSnapshot();
  });

  it("renders the first <Redirect> that matches the URL", () => {
    expect(
      render(
        <MemoryRouter initialEntries={["/three"]}>
          <Switch>
            <Route path="/one" render={() => <h1>one</h1>} />
            <Route path="/two" render={() => <h1>two</h1>} />
            <Redirect path="/three" to="/two" />
          </Switch>
        </MemoryRouter>
      ).asFragment()
    ).toMatchSnapshot();
  });

  it("does not render a second <Redirect> that also matches the URL", () => {
    expect(
      render(
        <MemoryRouter initialEntries={["/three"]}>
          <Switch>
            <Route path="/one" render={() => <h1>one</h1>} />
            <Route path="/two" render={() => <h1>two</h1>} />
            <Redirect path="/three" to="/two" />
            <Redirect path="/three" to="/one" />
          </Switch>
        </MemoryRouter>
      ).asFragment()
    ).toMatchSnapshot();
  });

  it("renders a Route with no `path` prop", () => {
    expect(
      render(
        <MemoryRouter initialEntries={["/two"]}>
          <Switch>
            <Route path="/one" render={() => <h1>one</h1>} />
            <Route render={() => <h1>two</h1>} />
          </Switch>
        </MemoryRouter>
      ).asFragment()
    ).toMatchSnapshot();
  });

  it("renders a Redirect with no `path` prop", () => {
    expect(
      render(
        <MemoryRouter initialEntries={["/three"]}>
          <Switch>
            <Route path="/one" render={() => <h1>one</h1>} />
            <Redirect to="/one" />
            <Route path="/two" render={() => <h1>two</h1>} />
          </Switch>
        </MemoryRouter>
      ).asFragment()
    ).toMatchSnapshot();
  });

  // it("handles subsequent redirects", () => {
  // const App = (
  // <MemoryRouter initialEntries={["/one"]}>
  // <Switch>
  // <Redirect path="/one" to="/two" />
  // <Redirect path="/two" to="/three" />
  // <Route path="/three" render={() => <h1>three</h1>} />
  // </Switch>
  // </MemoryRouter>
  // );

  // const { asFragment, rerender } = render(App);

  // expect(asFragment()).toMatchSnapshot();
  // rerender(App);
  // expect(asFragment()).toMatchSnapshot();
  // rerender(App);
  // expect(asFragment()).toMatchSnapshot();

  /// /waitForRedirects(() => {
  /// /expect(node.innerHTML).toContain("three");
  /// /done();
  /// /});
  // });

  it("handles comments", () => {
    expect(
      render(
        <MemoryRouter initialEntries={["/cupcakes"]}>
          <Switch>
            <Route path="/bubblegum" render={() => <div>bub</div>} />
            {/* this is a comment */}
            <Route path="/cupcakes" render={() => <div>cup</div>} />
          </Switch>
        </MemoryRouter>
      ).asFragment()
    ).toMatchSnapshot();
  });

  it("renders with non-element children", () => {
    expect(
      render(
        <MemoryRouter initialEntries={["/one"]}>
          <Switch>
            <Route path="/one" render={() => <h1>one</h1>} />
            {false}
            {undefined}
          </Switch>
        </MemoryRouter>
      ).asFragment()
    ).toMatchSnapshot();
  });

  it("can use a `location` prop instead of `router.location`", () => {
    expect(
      render(
        <MemoryRouter initialEntries={["/one"]}>
          <Switch location={{ pathname: "/two" }}>
            <Route path="/one" render={() => <h1>one</h1>} />
            <Route path="/two" render={() => <h1>two</h1>} />
          </Switch>
        </MemoryRouter>
      ).asFragment()
    ).toMatchSnapshot();
  });
});
