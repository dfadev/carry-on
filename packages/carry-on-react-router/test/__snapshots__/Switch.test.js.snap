// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A <Switch> can use a \`location\` prop instead of \`router.location\` 1`] = `
<DocumentFragment>
  <h1>
    two
  </h1>
</DocumentFragment>
`;

exports[`A <Switch> does not render a second <Redirect> that also matches the URL 1`] = `
<DocumentFragment>
  <h1>
    two
  </h1>
</DocumentFragment>
`;

exports[`A <Switch> does not render a second <Route> that also matches the URL 1`] = `
<DocumentFragment>
  <h1>
    one
  </h1>
</DocumentFragment>
`;

exports[`A <Switch> handles comments 1`] = `
<DocumentFragment>
  <div>
    cup
  </div>
</DocumentFragment>
`;

exports[`A <Switch> renders a Redirect with no \`path\` prop 1`] = `
<DocumentFragment>
  <h1>
    one
  </h1>
</DocumentFragment>
`;

exports[`A <Switch> renders a Route with no \`path\` prop 1`] = `
<DocumentFragment>
  <h1>
    two
  </h1>
</DocumentFragment>
`;

exports[`A <Switch> renders the first <Redirect> that matches the URL 1`] = `
<DocumentFragment>
  <h1>
    two
  </h1>
</DocumentFragment>
`;

exports[`A <Switch> renders the first <Route> that matches the URL 1`] = `
<DocumentFragment>
  <h1>
    one
  </h1>
</DocumentFragment>
`;

exports[`A <Switch> renders with non-element children 1`] = `
<DocumentFragment>
  <h1>
    one
  </h1>
</DocumentFragment>
`;
